store/types/store.go

```diff
@@ -38,6 +38,45 @@ type Queryable interface {
 //----------------------------------------
 // MultiStore
 
+type StoreUpgrades struct {
+	New     []string
+	Renamed []StoreRename
```

Comment:
This can just be a `map` from old to new esp given how `RenamedFrom` is written (I hate linear search!).

So if `New` isn't really used, why don't we simplify this whole thing and just use a `type StoreUpgrades map[string]string`, and a deletion just maps an old key to an empty string?

Reply:
I would not use maps anywhere near the DB.
If there are two renamed stores, then the order or renaming would be non-deterministic (thanks to golang map semantics).

This would mean a different write order in the iavl db, which can lead to a different tree structure, which leads to a different app hash and broken consensus.

I hit a similar issue in weave code when we stored indices in a map and updated in a non-deterministic order. After a few tx with that indexed fields, consensus halted.

Reply:
I like to make deletes specific

I will add json tags to this struct as well, as that is the plan.

Reply:
Okay that's fine, we can keep avoid the map, although I did notice that you were iterating over the story keys not the renames so probably wouldn't be an issue.

The bigger question is - why are developers not more enraged over the lack of basic data structures and algorithms in the golang stdlib?

Reply:
> Okay that's fine, we can keep avoid the map, although I did notice that you were iterating over the story keys not the renames so probably wouldn't be an issue.

Ah, actually I think you are right. But after being bit, I am very skeptical of `maps` anywhere near a `range`. And yes, go missed a lot of basic data structure algorithms. But this is due directly to the lack of generics. Only good one is `sort`.

`items.find(x => x.oldKey == "foo")` type work is impossible in go. Gotta learn to love cut and paste, and work on that typing speed